packages/flutter/test/gestures/long_press_test.dart

```diff
@@ -701,4 +701,45 @@ void main() {
     longPress.dispose();
     recognized.clear();
   });
+
+  testGesture('Switching buttons mid-stream does not fail to send "end" event', (GestureTester tester) {
+    final List<String> recognized = <String>[];
+    final LongPressGestureRecognizer longPress = LongPressGestureRecognizer()
+      ..onLongPressStart = (LongPressStartDetails details) {
+        recognized.add('primaryStart');
+      }
+      ..onLongPressEnd = (LongPressEndDetails details) {
+        recognized.add('primaryEnd');
+      };
+
+    const PointerDownEvent down4 = PointerDownEvent(
+      pointer: 8,
+      position: Offset(10, 10),
+    );
+
+    const PointerMoveEvent move4 = PointerMoveEvent(
+      pointer: 8,
+      position: Offset(100, 200),
+      buttons: kPrimaryButton | kSecondaryButton,
+    );
+
+    const PointerUpEvent up4 = PointerUpEvent(
+      pointer: 8,
+      position: Offset(100, 200),
+      buttons: kSecondaryButton,
+    );
+
+    longPress.addPointer(down4);
+    tester.closeArena(4);
+    tester.route(down4);
+    tester.async.elapse(const Duration(milliseconds: 1000));
+    recognized.add('two seconds later...');
+    tester.route(move4);
```

Comment:
From testing on macOS, after this first `move4`, when releasing the primary button, another move event is dispatched with the buttons set as `2`.